### PR TITLE
New observer system

### PR DESCRIPTION
**Pull request by basine**
**Note:** This needs a new release of ZMCDataModel, WireRequestStrategy and WireMessageStrategy

**Changes in this PR**
1) Refactoring of the WireCallCenterV2
2) Adding and tearing down SearchUserObserverSnapshots when searchDirectory changes or is torndown
3) Adjusting tests using new observer system

**TODO**
- [x] merge & release with https://github.com/wireapp/wire-ios-data-model/pull/138
- [x] release WireRequestStrategy with new data model
- [x] merge & release with https://github.com/wireapp/wire-ios-message-strategy/pull/34
- [x] rerun all tests

### CODE REVIEW

**Comment by typfel** [comment]
How does the new observer system work when in the background, is it always reporting changes? `WireCallCenterV2` and everything build on top of it expects observers to also fire when in the background.

File: Source/Public/ZMUserSession.h [hunk 1]
```diff
@@ -32,7 +32,7 @@
 @class ZMConversation;
 @class UserClient;
 @class ZMProxyRequest;
-@class WireCallCenterV2;
+//@class WireCallCenterV2;
```
Comment (typfel) [inline Source/Public/ZMUserSession.h]: Delete

File: Source/Public/ZMUserSession.h [hunk 2]
```diff
@@ -133,7 +133,7 @@ extern NSString * const ZMTransportRequestLoopNotificationName;
 @property (nonatomic, readonly) TopConversationsDirectory *topConversationsDirectory;
 
 /// CallCenter for V2 calls
-@property (nonatomic, readonly) WireCallCenterV2 *wireCallCenterV2;
+//@property (nonatomic, readonly) WireCallCenterV2 *wireCallCenterV2;
```
Comment (typfel) [inline Source/Public/ZMUserSession.h]: Delete

File: Source/Synchronization/ZMSyncStrategy.m [hunk 3]
```diff
@@ -421,9 +426,7 @@ - (void)managedObjectContextDidSave:(NSNotification *)note;
         
         NSSet *conversationsWithCallChanges = [callStateChanges allContainedConversationsInContext:strongUiMoc];
         if (conversationsWithCallChanges != nil) {
-            [NSNotificationCenter.defaultCenter postNotificationName:WireCallCenterV2.CallStateDidChangeNotification
-                                                              object:nil
-                                                            userInfo:@{ @"updated": conversationsWithCallChanges }];
+            [strongUiMoc.wireCallCenterV2 callStateDidChangeWithConversations:conversationsWithCallChanges];
```
Comment (typfel) [inline Source/Synchronization/ZMSyncStrategy.m]: I think the plan was to move this into the generic userInfo merge.

`NSManagedObjectContext+UserInfoMerge.swift`

**Comment by basine** [comment]
@typfel It's currently not processing or forwarding any changes while in the background. I did a few tests with the app and it seemed to work fine nonetheless, but we might have to review this behaviour.

File: Source/Synchronization/ZMSyncStrategy.m (hunk 3 above)
Comment (basine) [inline Source/Synchronization/ZMSyncStrategy.m]: Maybe this should be a different PR

**Comment by typfel** [comment]
@basine I think we need to review it, you wouldn't be able to receive a call in the background.

**Comment by basine** [comment]
@typfel I tried this and it worked. Nevertheless this would require another ZMCDataModel release and should be a different PR in that case.

**Comment by typfel** [comment]
@basine well then

"It's currently not processing or forwarding any changes while in the background"

is not working as it should or you didn't try it Callkit  or with calling3.

**Comment by basine** [comment]
@typfel as I said, I tried calling while the app was in the background and it worked. We should merge this PR anyway and then fix the background behaviour, because this PR currently blocks the work for everyone else.

**Comment by typfel** [comment]
@basine right maybe we can do it an a follow up PR but we should be aware of what breaks in dev when we merge this. Need to disable internal and so on.

**Comment by basine** [comment]
@typfel I understand your concerns, but this is not going to be released right away anyway - I expect a few more bugs to be uncovered since the refactoring was rather major. Therefore there will be follow up PRs for sure.
